### ISSUE TITLE
Acceptable code rejected when -Werror -Wignore is used

### ISSUE THREAD

**Issue by nad** [agda/agda#3187]
Consider the following code:
```agda
open import Agda.Builtin.Bool

F : Bool → Set
F true = Bool
F _    = Bool
```
This code is accepted by Agda, but rejected when `-Werror -Wignore` is used:
```
Exact splitting is enabled, but the following clause could not be
preserved as definitional equalities in the translation to a case
tree:
  F _ = Bool
when checking the definition of F
```
Note that I did not turn on `--exact-split`. Adding `--no-exact-split` to the command line does not help.

**Comment by gallais** [comment]
It looks to me like [`applyFlagsToTCWarnings`](https://github.com/agda/agda/blob/be89d4a8b264dd2719cb8c601a2c7f45a95ba220/src/full/Agda/TypeChecking/Errors.hs#L233) is completely ignoring the `WarningMode`.

**Comment by gallais** [comment]
Another source of problems: when we are checking the main interface we ignore the flags:
https://github.com/agda/agda/blob/master/src/full/Agda/Interaction/Imports.hs#L817

Simply changing `IgnoreFlags` to `RespectFlags` there leads to an internal error in
[`Serialize.Instances.Internal`](https://github.com/agda/agda/blob/master/src/full/Agda/TypeChecking/Serialise/Instances/Internal.hs#L128) when checking [`Issue964`](https://github.com/agda/agda/blob/master/test/Succeed/Issue964.agda) which imports a file with
some unsolved metas.

It's not clear to me what this `IgnoreFlags` really ought to be. Should we only ignore
flags turning off the warnings related to unsolved metas?

**Comment by gallais** [comment]
Got it (I think)!

In `TypeChecking.Warnings`, [we classify](https://github.com/agda/agda/blob/master/src/full/Agda/TypeChecking/Warnings.hs#L106) `CoverageNoExactSplit` as an `ErrorWarning`.
These kind of warnings [cannot be bypassed by a user-set flag](https://github.com/agda/agda/blob/master/src/full/Agda/TypeChecking/Warnings.hs#L47).

I think it is a mistake to call the lack of an exact split a (non-fatal) error rather than just a
warning. I'll change that. And I'll try to see whether I get `applyFlagsToTCWarnings` up
to date too.

**Comment by jespercockx** [comment]
The reason why CoverageNoExactSplit is an ErrorWarning instead of a regular Warning is because it was originally only used when the user enabled `--exact-split` explicitly, in which case it makes sense to have a proper error. But now I guess it is better to have it as a warning, and use `-Werror` when necessary.

**Comment by nad** [comment]
Perhaps we shouldn't have both `--exact-split` and the corresponding warning. Can one enable `-Werror` for a subset of warnings, and still have other warnings enabled?

**Comment by gallais** [comment]
At the moment `Werror` sets a boolean telling Agda we want all warnings to be turned
into errors. I suppose we could have a more fine-grained approach and offer `-E(no)Flag`
as the counterpart of `-W(no)Flag` but for treating them as errors rather than warnings.

**Comment by nad** [comment]
I have another example that checks without problems without `-Werror`, but fails with `-Werror`. I'll try to construct a small test case later.

**Comment by nad** [comment]
The following code is rejected:
```agda
{-# OPTIONS -Werror #-}

open import Agda.Builtin.Nat

data ⊥ : Set where

record ⊤ : Set where

data _⊎_ (A B : Set) : Set where
  inj₁ : A → A ⊎ B
  inj₂ : B → A ⊎ B

Fin : Nat → Set
Fin zero    = ⊥
Fin (suc n) = ⊤ ⊎ Fin n

postulate
  P : Set → Set
  Q : Nat → Set
  F : ∀ {n} → P (Fin n) → Q n

G : P (Fin 48) → Q 48
G = F
```
```
Refusing to invert pattern matching of Fin because the maximum
depth (50) has been reached. Most likely this means you have an
unsatisfiable constraint, but it could also mean that you need to
increase the maximum depth using the flag --inversion-max-depth=N
when checking that the expression F has type P (Fin 48) → Q 48
```
However, if `-Werror` is not turned on, then the code is accepted, without any warnings.

**Comment by gallais** [comment]
Right. These warnings are [filtered out after the fact](https://github.com/agda/agda/blob/2b1fe5621af463217d190b60003b9229f0317bb9/src/full/Agda/Interaction/Imports.hs#L943) if the unifier manages to find a
solution despite inversion failing.

One way to handle that would be to:
* not fail when `-Werror` is set and a warning of interest is raised
* fail once we wrap-up (where we currently are filtering out `InversionDepthReached`).

But even without the `-Werror` bug, the logic used here looks a bit dodgy to me tbh:
if we still have unsolved metas, it seems we will display *all* of the `InversionDepthReached`
warnings rather than only the ones corresponding to the set of unsolved metas.

ping @UlfNorell

**Comment by UlfNorell** [comment]
It's not completely obvious what inversion depth warnings correspond to a particular meta. At the point we're issuing the warning we have a constraint `f us = v` where `us` contains some metas. One possibility would be to gather the metas in `us` and not issue the warning if they have all been completely solved.

**Comment by andreasabel** [comment]
Removing the label `exact split` because the problem with exact split warnings has been resolved.

Admonition: please do not reopen issues that are fixed and closed, even if a (closely) related issue still persists.  Please open a new issue then.
